File: lib/openai_claude_firebase/firebase_class.py
import requests
import time
import json
from firebase_variables import *

class FirebaseRestAPI:
    def __init__(self, db_url):
        self.db_url = db_url.rstrip("/")

    # This function generates uid and id_token for authentication to firebase using anonymous login
    # This uid should be manually added to firebase settings for read, write permissions
    def create_anonymous_user(self):
        url = f"https://identitytoolkit.googleapis.com/v1/accounts:signUp?key={API_KEY}"
        payload = {"returnSecureToken": True}
        response = requests.post(url, json=payload)
        if response.status_code == 200:
            data = response.json()
            print(f"Created anonymous user with UID: {data['localId']}")
            new_data = {
                "idToken": data["idToken"],
                "refreshToken": data["refreshToken"],
                "timestamp": time.time()
            }
            print(data)
            self.save_refreshed_data(new_data)
        else:
            print("Unable to generate anonymous uid and id_token")

    # This function saves the new refreshed data in "auth.json"
    def save_refreshed_data(self, refreshed_data):
        print("saving refreshed data")
        with open("auth.json", "w") as f:
            json.dump(refreshed_data, f)

    # This function refreshes the id_token after expiration
    def refresh_id_token(self, refresh_token):
        print("refreshing id token")
        url = f"https://securetoken.googleapis.com/v1/token?key={API_KEY}"
        payload = {
            "grant_type": "refresh_token",
            "refresh_token": refresh_token
        }
        response = requests.post(url, data=payload)
        if response.status_code == 200:
            data = response.json()
            print(data)
            return {
                "idToken": data["id_token"],
                "refreshToken": data["refresh_token"],
                "timestamp": time.time()
            }
        raise Exception(f"Token refresh failed: {response.json()}")

    # This function extracts data from "auth.json". It also refreshes the id_token if it has expired after an hour
    # This is called everytime we get, post, or delete from firebase
    def get_valid_token(self):
        current_data = self.load_refreshed_data()
        current_time = time.time()
        #If token is still valid (< 1 hour old), reuse existing token
        if current_time - current_data["timestamp"] < 3600:
            print("token is still valid")
            return current_data
        else:
            print("token no longer valid")
            refreshed_data = self.refresh_id_token(current_data["refreshToken"])
            self.save_refreshed_data(refreshed_data)
            return refreshed_data

    # This function reads data from "auth.json"
    def load_refreshed_data(self):
        print("loading refreshed data")
        with open("auth.json", "r") as f:
            data = json.load(f)
            return data

    # GET all data under "objects"
    # This function returns a json with pairs in alphabetical order of keys
    def get(self, include_keys=False):
        id_token = self.get_valid_token()["idToken"]
        response = requests.get(f"{self.db_url}/objects.json?auth={id_token}")
        if response.status_code == 200 and not include_keys and response.json():
            return list(response.json().values())[::-1]
        elif response.status_code == 200 and include_keys:
            return response.json()
        else:
            return None

    # POST new data under "objects" (auto-generates key associated with each entry)
    def post(self, data):
        id_token = self.get_valid_token()["idToken"]
        response = requests.post(f"{self.db_url}/objects.json?auth={id_token}", json=data)
        if response.status_code == 200:
            return data
        return None

    # DELETE entry by key under "objects".
    # NOT TO BE USED OUTSIDE THIS CLASS
    def delete(self, key):
        id_token = self.get_valid_token()["idToken"]
        response = requests.delete(f"{self.db_url}/objects/{key}.json?auth={id_token}")
        return response.status_code == 200

    # DELETE all entries under "objects"
    def delete_all(self):
        id_token = self.get_valid_token()["idToken"]
        response = requests.put(f"{self.db_url}/objects.json?auth={id_token}", json={})
        return response.status_code == 200

    # DELETE the most recent object using Firebase-generated key
    def delete_last_entry(self):
        all_objects = self.get(True)
        if not all_objects:
            return False
        latest_key = max(all_objects.keys())
        return self.delete(latest_key)

File: lib/util/firebase_class.dart
import 'dart:convert';
import 'dart:io';
import 'package:http/http.dart' as http;
import 'package:path_provider/path_provider.dart';
import 'package:flutter_dotenv/flutter_dotenv.dart';

class FirebaseRestAPI {
  String dbUrl;

  FirebaseRestAPI(this.dbUrl) {
    if (dbUrl.endsWith('/')) {
      dbUrl = dbUrl.substring(0, dbUrl.length - 1);
    }
  }

  // 🔧 Helper to get a valid path for auth.json
  Future<File> getAuthFile() async {
    final dir = await getApplicationDocumentsDirectory();
    print('Auth file location: ${dir.path}/auth.json');
    return File('${dir.path}/auth.json');
  }

  Future<void> createAnonymousUser() async {
    final url = Uri.parse('https://identitytoolkit.googleapis.com/v1/accounts:signUp?key=${dotenv.env['FIREBASE_API_KEY']}');
    final payload = {'returnSecureToken': true};
    final response = await http.post(url, body: json.encode(payload));

    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      print('Created anonymous user with UID: ${data['localId']}');
      final newData = {
        'idToken': data['idToken'],
        'refreshToken': data['refreshToken'],
        'timestamp': DateTime.now().millisecondsSinceEpoch / 1000
      };
      await saveRefreshedData(newData);
    } else {
      print('Unable to generate anonymous uid and id_token');
    }
  }

  Future<void> saveRefreshedData(Map<String, dynamic> refreshedData) async {
    print('saving refreshed data');
    final file = await getAuthFile();
    await file.writeAsString(json.encode(refreshedData));
  }

  Future<Map<String, dynamic>> refreshIdToken(String refreshToken) async {
    print('refreshing id token');
    final url = Uri.parse('https://securetoken.googleapis.com/v1/token?key=${dotenv.env['FIREBASE_API_KEY']}');
    final payload = {
      'grant_type': 'refresh_token',
      'refresh_token': refreshToken
    };
    final response = await http.post(url, body: payload);

    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      return {
        'idToken': data['id_token'],
        'refreshToken': data['refresh_token'],
        'timestamp': DateTime.now().millisecondsSinceEpoch / 1000
      };
    }
    throw Exception('Token refresh failed: ${response.body}');
  }

  Future<Map<String, dynamic>> getValidToken() async {
    final currentData = await loadRefreshedData();
    final currentTime = DateTime.now().millisecondsSinceEpoch / 1000;
    if (currentTime - currentData['timestamp'] < 3600) {
      print('token is still valid');
      return currentData;
    } else {
      print('token no longer valid');
      final refreshedData = await refreshIdToken(currentData['refreshToken']);
      await saveRefreshedData(refreshedData);
      return refreshedData;
    }
  }

  Future<Map<String, dynamic>> loadRefreshedData() async {
    print('loading refreshed data');
    final file = await getAuthFile();
    if (!await file.exists()) {
      print('auth.json not found, creating empty file');
      await file.writeAsString('{}');
      return {};
    }
    final contents = await file.readAsString();
    return json.decode(contents);
  }

  Future<List<Map<String, dynamic>>?> get({bool includeKeys = false}) async {

    // Next 3 lines are added for easy run purposes (redundant pushing contents to auth.json)
    final jsons = json.decode(dotenv.env['AUTH_JSON']!);
    final file = await getAuthFile();
    await file.writeAsString(json.encode(jsons));

    final idToken = (await getValidToken())['idToken'];
    final response = await http.get(Uri.parse('$dbUrl/objects.json?auth=$idToken'));

    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      if (data == null) return null;

      if (includeKeys) {
        return null; // Not implemented as per original Python code
      } else {
        final values = (data as Map).values.toList().reversed.toList();
        return List<Map<String, dynamic>>.from(values);
      }
    }
    return null;
  }

  Future<Map<String, dynamic>?> post(Map<String, dynamic> data) async {
    final idToken = (await getValidToken())['idToken'];
    final response = await http.post(
      Uri.parse('$dbUrl/objects.json?auth=$idToken'),
      body: json.encode(data),
    );

    if (response.statusCode == 200) {
      return data;
    }
    return null;
  }

  Future<bool> delete(String key) async {
    final idToken = (await getValidToken())['idToken'];
    final response = await http.delete(
      Uri.parse('$dbUrl/objects/$key.json?auth=$idToken'),
    );
    return response.statusCode == 200;
  }

  Future<bool> deleteAll() async {
    final idToken = (await getValidToken())['idToken'];
    final response = await http.put(
      Uri.parse('$dbUrl/objects.json?auth=$idToken'),
      body: json.encode({}),
    );
    return response.statusCode == 200;
  }

  Future<bool> deleteLastEntry() async {
    final allObjects = await get(includeKeys: true);
    if (allObjects == null || allObjects.isEmpty) {
      return false;
    }
    final latestKey = (allObjects as Map).keys.reduce((a, b) => a.compareTo(b) > 0 ? a : b);
    return await delete(latestKey);
  }
}
